sling/myelin/express.cc

```diff
@@ -811,6 +811,143 @@ int Express::Complexity() const {
   return n;
 }
 
+bool Express::SparseCompatible() const {
+  for (Var *v : vars_) {
+    if (v->type == OUTPUT && !ZeroFixpoint(v)) return false;
+  }
+  return true;
+}
+
+bool Express::SparseAssignCompatible() const {
+  Var *in = Lookup(INPUT, 0);
+  Var *out = Lookup(OUTPUT, 0);
+  if (in == nullptr || out == nullptr) return false;
+  return Additive(out, in) && SparseCompatible();
+}
+
+bool Express::ZeroFixpoint(Var *x) const {
+  Op *op = x->producer;
+  if (op == nullptr) {
+    if (x->type == NUMBER) return x->id == ZERO;
+    return !x->single;
+  }
+  switch (op->type) {
+    case MOV:
+      return ZeroFixpoint(op->args[0]);
+    case ADD:
+    case SUB:
+    case MINIMUM:
+    case MAXIMUM:
+    case AND:
+    case BITAND:
+      return ZeroFixpoint(op->args[0]) && ZeroFixpoint(op->args[1]);
+    case MUL:
+    case OR:
+    case BITOR:
+      return ZeroFixpoint(op->args[0]) || ZeroFixpoint(op->args[1]);
+    case DIV:
+    case POW:
+      return ZeroFixpoint(op->args[0]);
+    case NEG: case ABS: case SIGN: case RELU: case SOFTSIGN:
+    case SQUARE: case SQRT: case SIGMOID: case ERF:
+    case SIN: case TAN:
+    case ASIN: case ATAN:
+    case SINH: case TANH:
+    case ASINH: case ATANH:
+    case FLOOR: case CEIL: case ROUND: case TRUNC:
+    case SUM:
+      return ZeroFixpoint(op->args[0]);
+    default:
+      return false;
+  }
+}
+
+bool Express::AlwaysZero(Var *x) const {
+  if (x->type == NUMBER && x->id == ZERO) return true;
+  Op *op = x->producer;
+  if (op == nullptr) return false;
+  switch (op->type) {
+    case MOV:
+      return AlwaysZero(op->args[0]);
+    case ADD:
+    case SUB:
+      return AlwaysZero(op->args[0]) && AlwaysZero(op->args[1]);
+    case MUL:
+      return AlwaysZero(op->args[0]) || AlwaysZero(op->args[1]);
+    case DIV:
+      return AlwaysZero(op->args[0]);
+    default:
+      return false;
+  }
+}
+
+bool Express::AlwaysOne(Var *x) const {
+  if (x->type == NUMBER && x->id == ONE) return true;
+  Op *op = x->producer;
+  if (op == nullptr) return false;
+  switch (op->type) {
+    case MOV:
+      return AlwaysOne(op->args[0]);
+    case ADD:
+      return (AlwaysOne(op->args[0]) && AlwaysZero(op->args[1])) ||
+             (AlwaysZero(op->args[0]) && AlwaysOne(op->args[1]));
+    case SUB:
+      return AlwaysOne(op->args[0]) && AlwaysZero(op->args[1]);
+    case MUL:
+    case DIV:
+      return AlwaysOne(op->args[0]) && AlwaysOne(op->args[1]);
+    default:
+      return false;
```

Comment:
Could these be added?
case ABS: case SQUARE: case SQRT: case FLOOR: case CEIL: case ROUND: case TRUNC:
  AlwaysOne(op->args[0])

Reply:
Again, they are not always one.